validator: add `set-tpu` command

diman-io opened the pull request:
#### Problem

Only one way to change this now is to restart a validator.
Also it could prove useful for the watchdog in the event of a remote service failure.

Also I have a question about tpu-forward #30451 

#### Summary of Changes

Add `set-tpu` subcommand to `solana-validotor`.

joncinque commented:
This overall seems reasonable to me, but to be honest, I'm not an expert in this area.

The one issue I see is the invalidation of the previous TLS certificate created at startup at https://github.com/solana-labs/solana/blob/d2fc2e5f62a5415ddbfdea8030b576d34339df13/streamer/src/quic.rs#L61 -- @pgarg66 will this cause other issues later on? Does this new endpoint need to reissue the certificate as well?

pgarg66 commented:
The QUIC connection cache currently fills in the IP address, but it is unused for connection lookup on the server side. That being said, it'll be best to update the client cert to make it future proof.

This new argument will we available for a non RPC (full) node as well. But looks like it will not actually set the TPU address. I find it a bit confusing.

diman-io commented:
This argument isn't new. There's a `--tpu-host-addr` option. I only add the subcommand to change this value on the fly.

This TPU address is only used to broadcast it through gossip. Other nodes use it to send packets to, kind of like a balancer before the node. A listen service on this address has to forward those packets to the node, possibly after performing some work on them. I didn't think it needs a cert for this purpose.

pgarg66 commented:
> This argument isn't new. There's a `--tpu-host-addr` option. I only add the subcommand to change this value on the fly.
> 
> This TPU address is only used to broadcast it through gossip. Other nodes use it to send packets to, kind of like a balancer before the node. A listen service on this address has to forward those packets to the node, possibly after performing some work on them. I didn't think it needs a cert for this purpose.

The node binds to the address provided via `--tpu-host-addr` option. This new option, as you also mentioned, is only used by the gossip broadcast to update the TPU address. The name of the option is kind of misleading. For example, a node operator (with no balancer) can set this value assuming the TPU will bind to it.

I agree that the QUIC cert doesn't need to be updated for the balancer case. But, the option name, description and details must be updated to prevent someone from accidentally misconfiguring their node.

pgarg66 commented:
> I was surprised to read this, but the node doesn't actually bind to the `--tpu-host-addr` option, but rather only uses it in gossip, which is why this change should be OK.
> 
Thanks @joncinque for looking into it, looks like I misread the code regarding this.

pgarg66 commented:
@diman-io, @joncinque and I had a quick sync up on this. So, the issue is that the IP address in the QUIC server cert uses the TPU IP address from the gossip record. Since this option changes the IP in gossip, the value in the cert will not match the IP to with the TPU is actually binding.

Currently the IP field in the cert is unused. But it could change in future. So for consistency sake, we should fill the correct IP in the cert.

I'll file an issue and we can handle it in a separate PR. This PR should be good to go.

joncinque commented on gossip/src/cluster_info.rs:
```diff
@@ -671,6 +671,17 @@ impl ClusterInfo {
+        let _ = self.my_contact_info.write().unwrap().set_tpu(socket);
+
+        self.insert_self();
+        self.push_message(CrdsValue::new_signed(
+            CrdsData::Version(Version::new(self.id())),
+            &self.keypair(),
+        ));
```
Is this extra message needed? It seems like you should just need an `insert_self()` and `push_self()` to get the contact info out.

@behzadnouri can you confirm if this message is needed to propagate a change to just the tpu host address?

diman-io replied:
Yes, you are right, my bad, careless copypasta

joncinque reviewed (commented):
This is looking close, just a few bits to clean up

behzadnouri commented on gossip/src/cluster_info.rs:
```diff
@@ -671,6 +671,13 @@ impl ClusterInfo {
     }
 
+    pub fn set_tpu(&self, tpu_addr: SocketAddr) {
+        let _ = self.my_contact_info.write().unwrap().set_tpu(tpu_addr);
```
can't ignore the error here

diman-io replied:
@behzadnouri Could you please check commit [1f17f4f](https://github.com/solana-labs/solana/pull/30452/commits/1f17f4f523429866ef253ec18c497d533588e52a)? I hope I understood you correctly.

behzadnouri commented:
I am confused what this actually do.
You change the tpu address to some other address in contact-info.
Isn't there any thread listening on that address that should be updated to listen to the new address?

diman-io commented:
> I am confused what this actually do.
> You change the tpu address to some other address in contact-info.
> Isn't there any thread listening on that address that should be updated to listen to the new address?

The validator always listens on its "default" TPU port. However, by using the `--set-tpu-addr` option, it's possible to specify a different TPU address (even on a different host) when starting the validator. This alternate address will be broadcast to the gossip. The service is expected to forward these packets to the validator after performing certain tasks, such as deduplication for example. The Jito client also works by receiving this traffic instead of the validator.

This pull request adds the capability to change the TPU address while the validator is running, without needing to restart it.

behzadnouri commented:
> The validator always listens on its "default" TPU port. However, by using the --set-tpu-addr option, it's possible to specify a different TPU address (even on a different host) when starting the validator. This alternate address will be broadcast to the gossip. The service is expected to forward these packets to the validator after performing certain tasks, such as deduplication for example. The Jito client also works by receiving this traffic instead of the validator.

Ok, then this is going to be pretty confusing. Can you put above as comments where needed in the code.
And also change to `--set-public-tpu-addr` and other places from `set_tpu_...` to `set_public_tpu_...`, etc

diman-io commented:
> Ok, then this is going to be pretty confusing. Can you put above as comments where needed in the code. And also change to `--set-public-tpu-addr` and other places from `set_tpu_...` to `set_public_tpu_...`, etc

I looked and found that the validator has the following address-related arguments:
```
--bind-address <HOST>
--public-rpc-address HOST:PORT
--rpc-bind-address <HOST>
--rpc-faucet-address HOST:PORT
--tpu-host-addr HOST:PORT
```
Early in this PR I already changed `tpu-host-addr` to `tpu-host-address`. Now I changed it to `public-tpu-address` and removed `host` from the name. Similarly, I changed the new command to `set-public-tpu-address` and all the new code. I also renamed `overwrite_tpu_addr` to `public_tpu_addr` in a pair places. You were right, the code is now more understandable.

As for the explanations, I think the ones in the validator's help are informative enough:
```
--public-tpu-address <HOST:PORT>
            Specify TPU address to advertise in gossip [default: ask --entrypoint or localhostwhen --entrypoint is not
            provided]
```
and
```
set-public-tpu-address     Specify TPU address to advertise in gossip
```

@behzadnouri Could you please review?

behzadnouri commented on validator/src/admin_rpc_service.rs:
```diff
@@ -588,6 +595,35 @@ impl AdminRpc for AdminRpcImpl {
+        public_tpu_addr: SocketAddr,
+    ) -> Result<()> {
+        debug!(
+            "set_public_tpu_address rpc request received: {:?}",
+            public_tpu_addr
```
https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args

similarly other place below.

behzadnouri commented on validator/src/cli.rs:
```diff
@@ -372,8 +372,9 @@ pub fn app<'a>(version: &'a str, default_args: &'a DefaultArgs) -> App<'a, 'a> {
         .arg(
-            Arg::with_name("tpu_host_addr")
-                .long("tpu-host-addr")
+            Arg::with_name("public_tpu_addr")
+                .long("public-tpu-address")
+                .alias("tpu-host-addr")
```
do we actually need to change this in this PR?
seems like a separate change, and unrelated to adding the admin rpc command.

diman-io replied:
@behzadnouri
No, we do not.
Initially, the [suggestion](https://github.com/solana-labs/solana/pull/30452#discussion_r1131851386) was made by @joncinque to rename the argument from `tpu-host-addr` to `tpu-host-address`, similar to the new subcommand `set-tpu-host-address`. 
Then, it was [suggested](https://github.com/solana-labs/solana/pull/30452#issuecomment-1475316593) by you to rename the subcommand to `set-public-tpu-addr`.
Following [some research](https://github.com/solana-labs/solana/pull/30452#issuecomment-1490188396), I renamed it to `set-public-tpu-address`, and as the argument was previously changed in this PR, I also changed it to `public-tpu-address`.

So, I will revert `tpu-host-addr` renaming and create a separate PR

diman-io commented on validator/src/admin_rpc_service.rs:
```diff
@@ -588,6 +595,29 @@ impl AdminRpc for AdminRpcImpl {
+            warn!(
+                "Public TPU address set to {}",
+                post_init.cluster_info.my_contact_info().tpu().unwrap()
+            );
```
Done
